### ISSUE TITLE
Question on Experimental Reproduction on CIFAR10-LT reverse settings

### ISSUE THREAD

**Issue by kevin04087** [zhouzihao11/CCL#1]
First of all, I would like to express my admiration for your excellent research. I have been reproducing the results from your paper, and while the reported accuracy on CIFAR10-LT under the reverse setting with N=500 and M=4000 is 85, I obtained around 82.3 when running the experiments with exactly the same code and default settings. Even considering randomness from different seeds, this gap seems larger than what I would expect from normal variation. I suspect that I might have made a mistake in running the experiments, but I was wondering if there might be an issue in my setup or perhaps something in the code itself. I would greatly appreciate it if you could clarify this.

**Comment by zhouzihao11** [comment]
Thank you very much for your kind words and for taking the time to reproduce our work. I truly appreciate your detailed feedback.

You've raised an important point. We have identified that the issue you encountered was indeed due to a problem in our current codebase. The root cause was an instability in the distribution estimation process. Specifically, the estimator was updated only once every 500 iterations, which, combined with an excessively large update magnitude, could lead to high variance in performance under certain settings, exactly as you observed.

This issue has been resolved in the latest commit. The fix involves implementing a smoother estimation process with more frequent updates and a reduced update amplitude, which significantly stabilizes the training and improves reproducibility.

<img width="1104" height="113" alt="Image" src="https://github.com/user-attachments/assets/3376e4f5-0806-42da-9414-ebb6e8c25929" />

To obtain stable and consistent results matching the paper, please pull the latest version of the code. You can then run the experiment using the following command, which is configured to apply the corrected smoothing method:

> python train.py --dataset cifar10 --num-max 500 --num-max-u 4000 --arch wideresnet --batch-size 64 --lr 0.03 --seed 0 --imb-ratio-label 100 --imb-ratio-unlabel 100 --out out/cifar-10/N500_M4000/reverse_100 --beta 0.999 --alpha 0.999 --est-epoch 5 --upbatch 100 --gpu-id 0 --flag-reverse-LT 1
>  

This command should yield results much closer to the reported  accuracy.

<img width="699" height="114" alt="Image" src="https://github.com/user-attachments/assets/e3f9d190-d156-4d46-932f-1366a0d352cd" />

[log.txt](https://github.com/user-attachments/files/21928523/log.txt)

 We apologize for any inconvenience this may have caused, and we are very grateful for your question, which helps improve the clarity and reliability of our work.

Please don't hesitate to let me know if you have any more questions or if you continue to experience issues.

**Comment by kevin04087** [comment]
Thank you very much for your kind response.
In the reverse setting, I can only reproduce the reported performance if I modify the computation of
logits_la_u = (- compute_adjustment_by_py((1 - pro) * py_labeled + pro * py_all, 1.0, args)
by replacing the value 1.0 with 2.0. Conversely, in the consistent setting, this modification prevents reproduction, whereas keeping the value as 1.0 works—but then the reverse setting still fails to reproduce.

Based on this, my understanding is that the appropriate value needs to be adjusted differently depending on the setting.
I was wondering if this understanding is correct.
Thank you again!

**Comment by zhouzihao11** [comment]
Thank you for your detailed observation and feedback!  

You are correct that the value previously needed adjustment depending on the setting. However, we have now unified the parameter to 1.0 for both the reverse and consistent settings. This change ensures that the reported performance can be reproduced consistently across different configurations.  

The corresponding fix has been updated in the repository. Please pull the latest version and try again with the value set to 1.0. You should now be able to reproduce the results in both settings.  Here is example results of set the value as 1.0 in reverse setting.

<img width="841" height="216" alt="Image" src="https://github.com/user-attachments/assets/a004456e-6964-4e89-9684-db3a438655bc" />

[log.txt](https://github.com/user-attachments/files/21990174/log.txt)

Let us know if you encounter any further issues—we appreciate your help in improving the implementation!

**Comment by zhouzihao11** [comment]
Thank you for your detailed observation and feedback!

As my points, performance not be particularly sensitive to this parameter. We have now unified the parameter to 1.0 for both the reverse and consistent settings.

The corresponding fix has been updated in the repository. Please pull the latest version and try again with the value set to 1.0. You should now be able to reproduce the results in both settings. Here is a example in reverse setting when set this value as 1.0

<img width="841" height="216" alt="Image" src="https://github.com/user-attachments/assets/461d9c24-d91b-401f-b79c-dc5693d61249" />

[log.txt](https://github.com/user-attachments/files/21990397/log.txt)

Let us know if you encounter any further issues—we appreciate your help in improving the implementation!

**Comment by kevin04087** [comment]
I re-ran the code and it now reproduces without any issues. Thank you very much for your help.

I have one more question. In computing targets_co, the adjustment uses logits_la_u, but I’m not sure which distribution this is actually estimating. Meanwhile, in logits_la_u_b, the adjustment is applied with respect to py_all. Since multiple distributions appear in different steps, I find it a bit confusing and would appreciate clarification.

Thank you again!

**Comment by zhouzihao11** [comment]
Thank you for the excellent question and insightful observation! You've perfectly captured the core design logic. Here's a clarification addressing your points, and why the two-step adjustment exists (while acknowledging your redundancy hypothesis):

Clarification of Distributions & Adjustments
1. ​​Unlabeled Loader Composition​ (In Code Implementation)​:

Exactly correct. As implemented (following prior work), it ​​combines labeled and unlabeled data​​.
→ py_all = Estimated ​​mixed distribution​​ (labeled + unlabeled).
→ py_unlabeled = Estimated ​​true unlabeled distribution​​.

2. Why Two Adjustments of logits_la_u?​

​​Step 1 (Post-hoc LA for Standard Head):
​
 ​  Why?​​ The standard classification head was trained ​​on the mixed loader​​.
   Goal:​​ Correct the head's bias toward the ​​mixed data's distribution​​ (py_all).

Step 2 (LA for Pseudo-labeling True Unlabeled Data):

  Why?​​ Now we want to generate pseudo-labels for the ​​true unlabeled data​​.
  Goal:​​ Adjust the already debiased logits (logits_la_u) to reflect the ​​target distribution​​ of purely unlabeled data (py_unlabeled).

3.  Why Adjustments of logits_la_u_b using py_all?

   In the learning process, all the mixed data are used for learning at the same time, so the learning of the balanced branch needs to use the mixed distribution for correction.


From my current point of view, such a design may be redundant, and the unification of py_all and py_unlabeled may not greatly affect the performance. Your deep dive into the implementation is impressive! We welcome further discussion or test results if you try the simplification.

**Comment by kevin04087** [comment]
Apologies for the delayed response. Your explanation has resolved my questions. I have not yet considered attempting the simplification, but I will explore possible ideas in the future and share them with you if I could contribute to your research. Thank you once again for your kind and detailed clarification.
